### ISSUE TITLE
Messages from fatalError/etc are no longer propagated on Bugsnag reports

### ISSUE THREAD

**Issue by edenman** [bugsnag/bugsnag-cocoa#318]
### Expected behavior
Forcing the app to crash by calling `fatalError("i am a message")` includes the "i am a message" text in the bugsnag report.

### Observed behavior
No message is included:
https://app.bugsnag.com/recharge-labs/ios-consumer/errors/5bbe93a371c2fc001a3e9ba8

### Version
Bugsnag 5.17.0
Xcode Version 10.0 (10A255)

### Additional information
Original feature request (#159) and PR to add the feature (#188).  I'm guessing that something in Swift 4.2 or Xcode 10 changed the way those notable addresses are attached to the crash report?  Or maybe broke the heuristics in some other way?

**Comment by edenman** [comment]
I just updated to Bugsnag 5.17.1 and the issue is still not fixed.  I tried both on a dev build and an ad-hoc distribution, no luck getting the fatalError message on either:

dev build https://app.bugsnag.com/recharge-labs/ios-consumer/errors/5c05587150c99b001aa35ac0
production https://app.bugsnag.com/recharge-labs/ios-consumer/errors/5c055ca9638ce3001a10ce8d

**Comment by snmaynard** [comment]
"i am a message" is less than 16 characters and is no longer possible to show. If you have a message of more than 16 characters it should work

**Comment by edenman** [comment]
Same issue, even with a message longer than 16 characters ("Eric test crash..") https://app.bugsnag.com/recharge-labs/ios-consumer/errors/5c0560f209f364001a4947be

**Comment by kattrali** [comment]
@edenman What device? I can try to reproduce and figure out what's going on.

**Comment by edenman** [comment]
@kattrali iPhone SE.  Swift 4.2, Xcode 10.0 (10A255)

**Comment by kattrali** [comment]
It seems to be working for me in a reduced test case using:

* iPhone SE + Swift 4.2 + Xcode 10.0 (10A255)
* iPhone SE + Swift 4.2 + Xcode 10.1 (10B61)

In both debug and production builds, and I tried a few different strings to rule out any weirdness.

I updated the test project [here](https://github.com/bugsnag/bugsnag-cocoa/tree/kattrali/upgrade-swift-ios-project/examples/swift-ios) (on the `kattrali/upgrade-swift-ios-project` branch) with the settings I'm using. Can you try it on your machine to see if it also works for you? If it does, perhaps we can see if a particular setting is causing different behavior.

![screenshot](https://user-images.githubusercontent.com/333454/49396897-15e8e000-f6ef-11e8-9867-54da80fa5bf9.png)

**Comment by edenman** [comment]
I just ran the test project, although I had to [tweak a few things](https://github.com/bugsnag/bugsnag-cocoa/compare/master...edenman:edenman/get-example-running) to get it working.  Do you see the crashes I forced?  I'm guessing they're being uploaded to a project that I don't have access to.

**Comment by kattrali** [comment]
By default there's no API key set, so you'd have to add one in the [app delegate](https://github.com/bugsnag/bugsnag-cocoa/blob/kattrali/upgrade-swift-ios-project/examples/swift-ios/bugsnag-example/AppDelegate.swift#L28).

**Comment by edenman** [comment]
Derp.  Ok, yeah, it's working fine in the sample app.  Not sure why it's not working in my actual app.

**Comment by sjmueller** [comment]
Can this be re-opened? I have same problem, anytime we invoke `fatalError("custom message")`, we get a blank message in bugsnag:

<img src="https://user-images.githubusercontent.com/29085/58914164-afab3d00-86d2-11e9-81b1-c19b8a1ae6f5.png" width="300" />

Our release builds to the App Store use bitcode, could that be affecting things?

**Comment by snmaynard** [comment]
@sjmueller `"custom message"` is less than 16 characters so would be known not to work. Are you actually trying with that string or a longer one?

**Comment by sjmueller** [comment]
Correct, we are experiencing this with more than 16 characters. Example: this is the line for `Actions.swift:742`:
```
guard routes.last != route.rawValue else { fatalError("Attempted to push the same route twice") }
```

**Comment by sjmueller** [comment]
Any update here? This is leaving us in the dark with our bug reports.

**Comment by edenman** [comment]
@sjmueller I wasn't able to figure out what was causing this.  I ended up just wrapping all `fatalError` calls in my own helper method that first adds a breadcrumb and then calls `fatalError`.  It sucks, but /shrug

**Comment by snmaynard** [comment]
The problem is with different versions of swift or the compiler that the memory location that this information in changes. I believe that we have to reach into specific registers to find it, and it depends on how long the string is as to where the information is. The reality is that swift needs to make an API for this so we can fetch it which I believe @kattrali is thinking about.

**Comment by sjmueller** [comment]
@snmaynard but that doesn't make too much sense if @edenman is writing the error to a breadcrumb before invoking the `fatalError`. If he's able to write the information to the breadcrumbs portion of the api, then why can't it be written to the right place automatically?

**Comment by snmaynard** [comment]
That's because he is doing it before calling fatalError. There is no API from a crash handler to get access to this information. We have to inspect registers to get it - and it's not reliable.

**Comment by sjmueller** [comment]
Thanks for the response. Do you have an alternative suggestion to invoking `fatalError` that would enable bugsnag to capture exceptions in these scenarios? In the case of `fatalError` it's our code invoking, so we could change to another pattern pretty quickly.

**Comment by snmaynard** [comment]
@kattrali may have some ideas - but what @edenman has done with wrapping it and adding a breadcrumb doesn't seem terrible to me. I'll see if I can think of other ideas - its annoying we cant get access to the data in the crash handler - I think @kattrali was looking to see if a proposal could be made to swift to change that.

**Comment by edenman** [comment]
My workaround works but it breaks grouping. And the issue I opened on that problem is stuck too: https://github.com/bugsnag/bugsnag-cocoa/issues/226

**Comment by Nemixdre** [comment]
Hi. Is there any news on that? It's been on hold for a while now.
Thanks :)

**Comment by tiwoc** [comment]
I just stumbled upon the same issue. The message we pass to `fatalError` is 18 characters long. There's a message included in the crash report, but its not the one we passed to `fatalError`.

To debug this, I took a look at the crashing binary with Hopper Disassembler and found that the error message that was passed to `fatalError` ("Test for Log.fatal") is stored at 0x1006c28f0, while the message that the crash handler included into the report ("al data for MSCB ") is located at 0x1006c28d0, which is just 32 (0x20) less than the correct address. There seems to be an error in the offset that the crash handler uses to determine the location of the message.

<img width="898" src="https://user-images.githubusercontent.com/1260821/75681227-71f43880-5c93-11ea-871d-0bd765b1b666.png">

This is in a simulator build for x86_64. Compiler is `11.0.0 (clang-1100.0.33.17)` from Xcode 11.3.1 (11C505).

**Comment by xander-jones** [comment]
Hey @Nemixdre, until we've got something more concrete from Swift, this isn't going to be progressed, unfortunately. As noted above, the locations for this data moves depending on a few factors, without a single and robust method for us to gather the data. We've made a proposal to Swift to make this change and are following any responses closely, we'll report back here as soon as there are any updates! :)

**Comment by sjmueller** [comment]
If we were to run Instabug in parallel, would they have the same exact problems? In other words, is this a Bugsnag issue or one that is unsolvable across the entire space?

**Comment by tiwoc** [comment]
For anyone wondering, the Swift proposal that was mentioned earlier in this thread is likely this one: [Cleanup callback for fatal Swift errors](https://forums.swift.org/t/stdlib-cleanup-callback-for-fatal-swift-errors/26977/21)

**Comment by sjmueller** [comment]
From that thread:

> The Swift compiler in development now emits fatal error traps with an artificial inline frame, which will allow crash reporters than understand debug info to show the error directly as part of the backtrace. Future toolchains that include this change should be able to do better.

Does this mean BugSnag has the ability to fix this now? If so, how long will it take?

**Comment by xander-jones** [comment]
Hey @sjmueller, it looks like this may now be possible. We'll take a look into this as priorities allow.

**Comment by sethfri** [comment]
Hi there, is there any update on this? This is something Crashlytics already does correctly, and we're disappointed to see Bugsnag doesn't

**Comment by johnkiely1** [comment]
Hi @sethfri,

We are working on this at the moment, although we don't have any definite timeframe as yet. We will share any updates here as we have them.

**Comment by mattdyoung** [comment]
We've now implemented a new approach to extracting the error messages from Swift's `assertionFailure`, `fatalError`, and `preconditionFailure` functions, which has been released in [v6.5.0](https://github.com/bugsnag/bugsnag-cocoa/releases/tag/v6.5.0).

For further details see:
https://github.com/bugsnag/bugsnag-cocoa/pull/948
